CompletionException: io.zeebe.journal.file.record.CorruptedLogException: Record doesn't match checksum. Log segment may be corrupted.

**Describe the bug**
Observed in logs of benchmark: https://console.cloud.google.com/errors/CI-Xjtayu4ONmQE?service=zeebe-broker&version=medic-cw-13-553958a93-benchmark&time=P7D&project=zeebe-io&folder&organizationId=669107107215

**Log/Stacktrace**
If possible add the full stacktrace or Zeebe log which contains the issue. 

<details><summary>Full Stacktrace</summary>
 <p>

```
java.util.concurrent.CompletionException: io.zeebe.journal.file.record.CorruptedLogException: Record doesn't match checksum. Log segment may be corrupted.
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1081) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478) ~[?:?]
	at io.atomix.utils.concurrent.SingleThreadContext$WrappedRunnable.run(SingleThreadContext.java:188) ~[atomix-utils-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) ~[?:?]
Caused by: io.zeebe.journal.file.record.CorruptedLogException: Record doesn't match checksum. Log segment may be corrupted.
	at io.zeebe.journal.file.record.JournalRecordReaderUtil.read(JournalRecordReaderUtil.java:68) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.reset(MappedJournalSegmentWriter.java:222) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.zeebe.journal.file.MappedJournalSegmentWriter.<init>(MappedJournalSegmentWriter.java:63) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.zeebe.journal.file.JournalSegment.createWriter(JournalSegment.java:153) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.zeebe.journal.file.JournalSegment.<init>(JournalSegment.java:57) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.zeebe.journal.file.SegmentedJournal.newSegment(SegmentedJournal.java:441) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.zeebe.journal.file.SegmentedJournal.loadSegment(SegmentedJournal.java:452) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.zeebe.journal.file.SegmentedJournal.loadSegments(SegmentedJournal.java:500) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.zeebe.journal.file.SegmentedJournal.open(SegmentedJournal.java:206) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.zeebe.journal.file.SegmentedJournal.<init>(SegmentedJournal.java:79) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.zeebe.journal.file.SegmentedJournalBuilder.build(SegmentedJournalBuilder.java:136) ~[zeebe-journal-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.raft.storage.log.RaftLog$Builder.build(RaftLog.java:321) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.raft.storage.RaftStorage.openLog(RaftStorage.java:242) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.raft.impl.RaftContext.<init>(RaftContext.java:155) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:257) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:231) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer(RaftPartitionServer.java:175) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.raft.partition.impl.RaftPartitionServer.initServer(RaftPartitionServer.java:145) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.raft.partition.impl.RaftPartitionServer.start(RaftPartitionServer.java:103) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.raft.partition.RaftPartition.open(RaftPartition.java:119) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.raft.partition.RaftPartitionGroup.lambda$join$6(RaftPartitionGroup.java:201) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1603) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
	at io.atomix.raft.partition.RaftPartitionGroup.join(RaftPartitionGroup.java:203) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.primitive.partition.impl.DefaultPartitionService.lambda$start$1(DefaultPartitionService.java:80) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.concurrent.ConcurrentHashMap$ValueSpliterator.forEachRemaining(ConcurrentHashMap.java:3605) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
	at io.atomix.primitive.partition.impl.DefaultPartitionService.start(DefaultPartitionService.java:81) ~[atomix-cluster-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.atomix.core.Atomix.lambda$startServices$0(Atomix.java:178) ~[atomix-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072) ~[?:?]
	... 8 more
```

</p>
</details>

<details><summary>Log (with Hidden Journal Segment logger for brevity)</summary>
 <p>

```

W 2021-03-31T09:44:25.151093Z RaftServer{raft-partition-partition-1} - AppendRequest{term=13, leader=1, prevLogIndex=202211362, prevLogTerm=12, entries=0, commitIndex=202369393} to 2 failed: java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:25.343930Z RaftServer{raft-partition-partition-1} - AppendRequest{term=13, leader=1, prevLogIndex=202211362, prevLogTerm=12, entries=0, commitIndex=202369824} to 2 failed: java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:25.529503Z RaftServer{raft-partition-partition-1} - AppendRequest{term=13, leader=1, prevLogIndex=202211362, prevLogTerm=12, entries=0, commitIndex=202370373} to 2 failed: java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:25.705240Z RaftServer{raft-partition-partition-1} - AppendRequest{term=13, leader=1, prevLogIndex=202211362, prevLogTerm=12, entries=0, commitIndex=202370803} to 2 failed: java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:25.878680Z RaftServer{raft-partition-partition-1} - AppendRequest{term=13, leader=1, prevLogIndex=202211362, prevLogTerm=12, entries=0, commitIndex=202371259} to 2 failed: java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:26.060699Z RaftServer{raft-partition-partition-1} - AppendRequest{term=13, leader=1, prevLogIndex=202211362, prevLogTerm=12, entries=0, commitIndex=202371615} to 2 failed: java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:26.277679Z RaftServer{raft-partition-partition-1} - AppendRequest{term=13, leader=1, prevLogIndex=202211362, prevLogTerm=12, entries=0, commitIndex=202372098} to 2 failed: java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:26.439233Z RaftServer{raft-partition-partition-1} - AppendRequest{term=13, leader=1, prevLogIndex=202211362, prevLogTerm=12, entries=0, commitIndex=202372359} to 2 failed: java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:26.602292Z RaftServer{raft-partition-partition-1} - AppendRequest{term=13, leader=1, prevLogIndex=202211362, prevLogTerm=12, entries=0, commitIndex=202372816} to 2 failed: java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:26.801878Z RaftServer{raft-partition-partition-1} - AppendRequest{term=13, leader=1, prevLogIndex=202211362, prevLogTerm=12, entries=0, commitIndex=202373206} to 2 failed: java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
I 2021-03-31T09:44:26.858439Z RaftServer{raft-partition-partition-1} - Transitioning to FOLLOWER 
I 2021-03-31T09:44:26.858511Z RaftPartitionServer{raft-partition-partition-3} - Starting server for partition PartitionId{id=3, group=raft-partition} 
I 2021-03-31T09:44:26.859415Z RaftServer{raft-partition-partition-1} - Server join completed. Waiting for the server to be READY 
I 2021-03-31T09:44:26.864600Z RaftServer{raft-partition-partition-1} - Found leader 1 
I 2021-03-31T09:44:26.890917Z RaftServer{raft-partition-partition-1} - Setting firstCommitIndex to 202373457. RaftServer is ready only after it has committed events upto this index 
I 2021-03-31T09:44:26.984315Z RaftServer{raft-partition-partition-1}{role=FOLLOWER} - Started receiving new snapshot FileBasedReceivedSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/pending/195037810-8-294441153-284558432-1, snapshotStore=io.zeebe.snapshots.broker.impl.FileBasedSnapshotStore@2a7408d7, metadata=FileBasedSnapshotMetadata{index=195037810, term=8, processedPosition=294441153, exporterPosition=284558432}} from 1 
D 2021-03-31T09:44:27.013775Z Consume snapshot snapshotChunk 023883.sst of snapshot 195037810-8-294441153-284558432 
D 2021-03-31T09:44:27.040415Z Consume snapshot snapshotChunk 023884.log of snapshot 195037810-8-294441153-284558432 
D 2021-03-31T09:44:27.168852Z Consume snapshot snapshotChunk 023885.sst of snapshot 195037810-8-294441153-284558432 
D 2021-03-31T09:44:27.180370Z Consume snapshot snapshotChunk CHECKSUM of snapshot 195037810-8-294441153-284558432 
D 2021-03-31T09:44:27.183806Z Consume snapshot snapshotChunk CURRENT of snapshot 195037810-8-294441153-284558432 
D 2021-03-31T09:44:27.188710Z Consume snapshot snapshotChunk MANIFEST-023745 of snapshot 195037810-8-294441153-284558432 
D 2021-03-31T09:44:27.192429Z Consume snapshot snapshotChunk OPTIONS-023748 of snapshot 195037810-8-294441153-284558432 
D 2021-03-31T09:44:27.211827Z Purging snapshots older than FileBasedSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/195037810-8-294441153-284558432, metadata=FileBasedSnapshotMetadata{index=195037810, term=8, processedPosition=294441153, exporterPosition=284558432}} 
D 2021-03-31T09:44:27.212709Z Deleting snapshot FileBasedSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/194352529-8-293328940-283552087, metadata=FileBasedSnapshotMetadata{index=194352529, term=8, processedPosition=293328940, exporterPosition=283552087}} 
D 2021-03-31T09:44:27.215680Z Search for orphaned snapshots below oldest valid snapshot with index FileBasedSnapshotMetadata{index=195037810, term=8, processedPosition=294441153, exporterPosition=284558432} in /usr/local/zeebe/data/raft-partition/partitions/1/pending 
D 2021-03-31T09:44:27.217723Z Created new snapshot FileBasedSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/195037810-8-294441153-284558432, metadata=FileBasedSnapshotMetadata{index=195037810, term=8, processedPosition=294441153, exporterPosition=284558432}} 
I 2021-03-31T09:44:27.218150Z RaftServer{raft-partition-partition-1}{role=FOLLOWER} - Committed snapshot FileBasedSnapshot{directory=/usr/local/zeebe/data/raft-partition/partitions/1/snapshots/195037810-8-294441153-284558432, metadata=FileBasedSnapshotMetadata{index=195037810, term=8, processedPosition=294441153, exporterPosition=284558432}} 
I 2021-03-31T09:44:27.218533Z RaftServer{raft-partition-partition-1}{role=FOLLOWER} - Delete existing log (lastIndex '195022930') and replace with received snapshot (index '195037810') 
I 2021-03-31T09:44:28.556204Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - No heartbeat from null in the last PT4.426S (calculated from last 4431 ms), sending poll requests 
W 2021-03-31T09:44:28.557125Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 2 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:28.557409Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 0 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
I 2021-03-31T09:44:33.102753Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - No heartbeat from null in the last PT4.545S (calculated from last 4545 ms), sending poll requests 
W 2021-03-31T09:44:33.103591Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 2 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:33.103828Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 0 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
I 2021-03-31T09:44:36.418079Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - No heartbeat from null in the last PT3.314S (calculated from last 3315 ms), sending poll requests 
W 2021-03-31T09:44:36.418837Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 2 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:36.419021Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 0 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
I 2021-03-31T09:44:40.789214Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - No heartbeat from null in the last PT4.37S (calculated from last 4370 ms), sending poll requests 
W 2021-03-31T09:44:40.790099Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 2 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:40.790346Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 0 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
I 2021-03-31T09:44:43.405649Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - No heartbeat from null in the last PT2.615S (calculated from last 2615 ms), sending poll requests 
W 2021-03-31T09:44:43.406721Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 2 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:43.406986Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 0 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
I 2021-03-31T09:44:46.158242Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - No heartbeat from null in the last PT2.751S (calculated from last 2751 ms), sending poll requests 
W 2021-03-31T09:44:46.159106Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 2 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:46.159296Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 0 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
I 2021-03-31T09:44:48.890507Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - No heartbeat from null in the last PT2.731S (calculated from last 2731 ms), sending poll requests 
W 2021-03-31T09:44:48.891299Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 2 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
W 2021-03-31T09:44:48.891676Z RaftServer{raft-partition-partition-3}{role=FOLLOWER} - Poll request to 0 failed: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message 
E 2021-03-31T09:44:52.561400Z Bootstrap Broker-2 [6/13]: cluster services failed with unexpected exception. 

```

</p>
</details>

**Environment:**
 - OS: [e.g. Linux]
 - Zeebe Version: [e.g. 0.20.0]
 - Configuration: [e.g. exporters etc.]

--- a/journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java
+++ b/journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java
@@ -19,6 +19,7 @@ package io.zeebe.journal.file;
 import io.zeebe.journal.JournalException.InvalidChecksum;
 import io.zeebe.journal.JournalException.InvalidIndex;
 import io.zeebe.journal.JournalRecord;
+import io.zeebe.journal.file.record.CorruptedLogException;
 import io.zeebe.journal.file.record.JournalRecordReaderUtil;
 import io.zeebe.journal.file.record.JournalRecordSerializer;
 import io.zeebe.journal.file.record.PersistedJournalRecord;
@@ -59,7 +60,7 @@ class MappedJournalSegmentWriter {
     firstIndex = segment.index();
     this.buffer = buffer;
     writeBuffer.wrap(buffer);
-    reset(0);
+    reset(0, true);
   }
 
   public long getLastIndex() {
@@ -188,11 +189,16 @@ class MappedJournalSegmentWriter {
   }
 
   private void reset(final long index) {
+    reset(index, false);
+  }
+
+  private void reset(final long index, final boolean startup) {
     long nextIndex = firstIndex;
 
     // Clear the buffer indexes.
     buffer.position(JournalSegmentDescriptor.BYTES);
     buffer.mark();
+    int position = buffer.position();
     try {
       while ((index == 0 || nextIndex <= index) && FrameUtil.readVersion(buffer).isPresent()) {
         final var nextEntry = recordUtil.read(buffer, nextIndex);
@@ -202,9 +208,19 @@ class MappedJournalSegmentWriter {
         lastEntry = nextEntry;
         nextIndex++;
         buffer.mark();
+        position = buffer.position();
       }
     } catch (final BufferUnderflowException e) {
       // Reached end of the segment
+    } catch (final CorruptedLogException e) {
+      // if at startup, assume record was partially written due to crash and mark as ignored
+      if (!startup) {
+        throw e;
+      }
+
+      FrameUtil.markAsIgnored(buffer, position);
+      buffer.position(position);
+      buffer.mark();
     } finally {
       buffer.reset();
     }

--- a/journal/src/test/java/io/zeebe/journal/file/JournalTest.java
+++ b/journal/src/test/java/io/zeebe/journal/file/JournalTest.java
@@ -23,10 +23,20 @@ import io.zeebe.journal.JournalException.InvalidChecksum;
 import io.zeebe.journal.JournalException.InvalidIndex;
 import io.zeebe.journal.JournalReader;
 import io.zeebe.journal.JournalRecord;
+import io.zeebe.journal.file.record.JournalRecordReaderUtil;
+import io.zeebe.journal.file.record.SBESerializer;
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Path;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Optional;
 import org.agrona.DirectBuffer;
 import org.agrona.concurrent.UnsafeBuffer;
 import org.junit.jupiter.api.BeforeEach;
@@ -36,6 +46,7 @@ import org.junit.jupiter.api.io.TempDir;
 class JournalTest {
 
   @TempDir Path directory;
+  @TempDir Path temp;
 
   private byte[] entry;
   private final DirectBuffer data = new UnsafeBuffer();
@@ -495,6 +506,102 @@ class JournalTest {
     assertThat(reader.next()).isEqualTo(firstRecord);
     assertThat(reader.next()).isEqualTo(secondRecord);
     assertThat(reader.hasNext()).isFalse();
+  }
+
+  @Test
+  public void shouldTruncateCorruptedEntriesOnStartup() throws Exception {
+    // given
+    data.wrap("000".getBytes(StandardCharsets.UTF_8));
+    final var firstRecord = journal.append(data);
+    final var secondRecord = journal.append(data);
+    assertThat(directory.toFile().listFiles()).hasSize(1);
+    assertThat(directory.toFile().listFiles()[0].listFiles()).hasSize(1);
+    final File log = directory.toFile().listFiles()[0].listFiles()[0];
+
+    final File tempFile = new File(temp.toFile(), "rand");
+    assertThat(tempFile.createNewFile()).isTrue();
+
+    // when
+    journal.close();
+    assertThat(corruptFile(tempFile, log, secondRecord.index())).isTrue();
+    journal = openJournal();
+    final var reader = journal.openReader();
+
+    // then
+    assertThat(journal.getLastIndex()).isEqualTo(firstRecord.index());
+    assertThat(reader.seekToLast()).isEqualTo(firstRecord.index());
+    assertThat(reader.next()).isEqualTo(firstRecord);
+  }
+
+  @Test
+  public void shouldWriteEntriesAfterTruncatedCorruptedEntries() throws Exception {
+    // given
+    data.wrap("000".getBytes(StandardCharsets.UTF_8));
+    final var firstRecord = journal.append(data);
+    final var secondRecord = journal.append(data);
+    assertThat(directory.toFile().listFiles()).hasSize(1);
+    assertThat(directory.toFile().listFiles()[0].listFiles()).hasSize(1);
+    final File log = directory.toFile().listFiles()[0].listFiles()[0];
+
+    final File tempFile = new File(temp.toFile(), "rand");
+    assertThat(tempFile.createNewFile()).isTrue();
+
+    // when
+    journal.close();
+    assertThat(corruptFile(tempFile, log, secondRecord.index())).isTrue();
+    journal = openJournal();
+    data.wrap("111".getBytes(StandardCharsets.UTF_8));
+    final var lastRecord = journal.append(data);
+    final var reader = journal.openReader();
+
+    // then
+    assertThat(reader.next()).isEqualTo(firstRecord);
+    assertThat(reader.next()).isEqualTo(lastRecord);
+  }
+
+  static boolean corruptFile(final File temp, final File file, final long index)
+      throws IOException {
+    final byte[] bytes = new byte[1024];
+    int read = 0;
+
+    try (final BufferedInputStream in = new BufferedInputStream(new FileInputStream(file))) {
+      while (in.available() > 0 && read < 1024) {
+        read += in.read(bytes, read, Math.min(1024, in.available()) - read);
+      }
+    }
+
+    if (!corruptRecord(bytes, index)) {
+      return false;
+    }
+
+    try (final BufferedOutputStream out = new BufferedOutputStream(new FileOutputStream(temp))) {
+      out.write(bytes, 0, read);
+    }
+
+    return temp.renameTo(file);
+  }
+
+  private static boolean corruptRecord(final byte[] bytes, final long targetIndex) {
+    final JournalRecordReaderUtil reader = new JournalRecordReaderUtil(new SBESerializer());
+    final ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    buffer.position(JournalSegmentDescriptor.BYTES);
+
+    Optional<Integer> version = FrameUtil.readVersion(buffer);
+    for (long index = 1; version.isPresent() && version.get() == 1; index++) {
+      final var record = reader.read(buffer, index);
+
+      if (record == null || record.index() > targetIndex) {
+        break;
+      } else if (record.index() == targetIndex) {
+        final int lastPos = buffer.position() - 1;
+        buffer.put(lastPos, (byte) ~buffer.get(lastPos));
+        return true;
+      }
+
+      version = FrameUtil.readVersion(buffer);
+    }
+
+    return false;
   }
 
   private SegmentedJournal openJournal() {

--- a/journal/src/test/java/io/zeebe/journal/file/JournalTest.java
+++ b/journal/src/test/java/io/zeebe/journal/file/JournalTest.java
@@ -24,6 +24,9 @@ import io.zeebe.journal.JournalException.InvalidIndex;
 import io.zeebe.journal.JournalReader;
 import io.zeebe.journal.JournalRecord;
 import io.zeebe.journal.file.record.JournalRecordReaderUtil;
+import io.zeebe.journal.file.record.PersistedJournalRecord;
+import io.zeebe.journal.file.record.RecordData;
+import io.zeebe.journal.file.record.RecordMetadata;
 import io.zeebe.journal.file.record.SBESerializer;
 import java.io.BufferedInputStream;
 import java.io.BufferedOutputStream;
@@ -426,7 +429,7 @@ class JournalTest {
   @Test
   void shouldReadReopenedJournal() throws Exception {
     // given
-    final var appendedRecord = journal.append(data);
+    final var appendedRecord = copyRecord(journal.append(data));
     journal.close();
 
     // when
@@ -442,7 +445,7 @@ class JournalTest {
   @Test
   void shouldWriteToReopenedJournalAtNextIndex() throws Exception {
     // given
-    final var firstRecord = journal.append(data);
+    final var firstRecord = copyRecord(journal.append(data));
     journal.close();
 
     // when
@@ -489,14 +492,15 @@ class JournalTest {
   public void shouldInvalidateAllEntries() throws Exception {
     // given
     data.wrap("000".getBytes(StandardCharsets.UTF_8));
-    final var firstRecord = journal.append(data);
+    final var firstRecord = copyRecord(journal.append(data));
+
     journal.append(data);
     journal.append(data);
 
     // when
     journal.deleteAfter(firstRecord.index());
     data.wrap("111".getBytes(StandardCharsets.UTF_8));
-    final var secondRecord = journal.append(data);
+    final var secondRecord = copyRecord(journal.append(data));
 
     journal.close();
     journal = openJournal();
@@ -512,8 +516,8 @@ class JournalTest {
   public void shouldTruncateCorruptedEntriesOnStartup() throws Exception {
     // given
     data.wrap("000".getBytes(StandardCharsets.UTF_8));
-    final var firstRecord = journal.append(data);
-    final var secondRecord = journal.append(data);
+    final var firstRecord = copyRecord(journal.append(data));
+    final var secondRecord = copyRecord(journal.append(data));
     assertThat(directory.toFile().listFiles()).hasSize(1);
     assertThat(directory.toFile().listFiles()[0].listFiles()).hasSize(1);
     final File log = directory.toFile().listFiles()[0].listFiles()[0];
@@ -537,8 +541,8 @@ class JournalTest {
   public void shouldWriteEntriesAfterTruncatedCorruptedEntries() throws Exception {
     // given
     data.wrap("000".getBytes(StandardCharsets.UTF_8));
-    final var firstRecord = journal.append(data);
-    final var secondRecord = journal.append(data);
+    final var firstRecord = copyRecord(journal.append(data));
+    final var secondRecord = copyRecord(journal.append(data));
     assertThat(directory.toFile().listFiles()).hasSize(1);
     assertThat(directory.toFile().listFiles()[0].listFiles()).hasSize(1);
     final File log = directory.toFile().listFiles()[0].listFiles()[0];
@@ -602,6 +606,18 @@ class JournalTest {
     }
 
     return false;
+  }
+
+  private PersistedJournalRecord copyRecord(final JournalRecord record) {
+    final DirectBuffer data = record.data();
+    final byte[] buffer = new byte[data.capacity()];
+    data.getBytes(0, buffer);
+
+    final UnsafeBuffer copiedData = new UnsafeBuffer(buffer);
+    final RecordData copiedRecord = new RecordData(record.index(), record.asqn(), copiedData);
+
+    return new PersistedJournalRecord(
+        new RecordMetadata(record.checksum(), copiedRecord.data().capacity()), copiedRecord);
   }
 
   private SegmentedJournal openJournal() {

